schema/project.js

```diff
@@ -69,6 +70,76 @@ class ProjectQueryBuilder extends QueryBuilder {
     );
   }
 
+  whereRopsDue(year) {
+    return this.where('projects.issueDate', '<=', `${year}-12-31`)
+      .andWhere(builder => {
+        builder
+          .where('projects.status', 'active')
+          .orWhere(qb => {
+            qb
+              .where('projects.status', 'expired')
+              .where('projects.expiryDate', '>=', `${year}-01-01`);
+          })
+          .orWhere(qb => {
+            qb
+              .where('projects.status', 'revoked')
+              .where('projects.revocationDate', '>=', `${year}-01-01`);
+          });
+      });
+  }
+
+  whereRopsSubmitted(year) {
+    return this.whereExists(
+      Project.relatedQuery('rops')
+        .where('year', year)
+        .where('rops.status', 'submitted')
+    );
+  }
+
+  whereRopsOutstanding(year) {
+    return this.whereNotExists(
+      Project.relatedQuery('rops')
+        .where('year', year)
+        .where('rops.status', 'submitted')
+    );
+  }
+
+  withRops(year, ropsStatus) {
+    const query = this.withGraphFetched('rops(constrainRops)')
+      .modifiers({
+        constrainRops: builder => {
+          builder.where('rops.year', year);
+          if (ropsStatus === 'submitted') {
+            builder.where('rops.status', 'submitted');
+          }
+        }
+      });
+
+    if (ropsStatus === 'outstanding') {
+      const endOfJanNextYear = moment(`${parseInt(year, 10) + 1}-01-31`).endOf('day').toISOString();
+      const interval28Days = `INTERVAL '29 days - 1 millisecond'`;
+      query.select(this.knex().raw(`
+        CASE
+          WHEN projects.status = 'active' THEN LEAST('${endOfJanNextYear}'::timestamptz, date_trunc('day', projects.expiry_date) + ${interval28Days})
+          WHEN projects.status = 'expired' THEN DATE_TRUNC('day', projects.expiry_date) + ${interval28Days}
+          WHEN projects.status = 'revoked' THEN DATE_TRUNC  ('day', projects.revocation_date) + ${interval28Days}
```

Comment:
I know it probably doesn't affect the functionality, but the inconsistency between `date_trunc(...)`, `DATE_TRUNC(...)` and `DATE_TRUNC (...)` is irritating.